### ISSUE TITLE
[RLlib] - Check environment numerics during run time

### ISSUE THREAD

**Issue by simonsays1980** [ray-project/ray#31430]
### Description

Following the discussion in #30471 (specifically [this comment](https://github.com/ray-project/ray/issues/30471#issuecomment-1361775334)) and much checking over the last years, I would like to see some tools that simplify tracing an error back to its cause. One of such tools is checking for numerics (NaNs/Infs) in the environment during runtime. 

I admit that environment checking could be considered as solely the user's responsibility, but in my opinion the framework that RLlib offers for training RL agents on any kind of environment has grown so complex and nested that it is sometimes hard for users that are not that experienced with the source code to trace back an error. In summary I see three reasons why checking numerics in an environment might me advantageous to be done inside of RLlib: 

1. We face sometimes error in the environment at runtime that come from some very specific actions - we cannot test for all situation beforehand. It is easier to check then during runtime what happens. 
2. RLlib has grown to a large powerful framework that contains many nested classes and complex workflows: 
          a. Most end users are not that familiar with the source code that they could easily check the numerics within the environment. 
          b. Debugging is still complicated and often needs the user to change some settings (e.g. `local_mode`).
          c. With RLlib we want to actually lift the heavy-weight away from the user and let her apply RL. 
2. Numerics might be checked on the environment side - but even here users face the same problem that they might need to change source code they have not written. 

### Use case

Having an environment I can set some configuration parameters similar to `disable_env_checking` that check the numerics in an environment at runtime. 

The checking can be done via printing warnings or - if the user wants it - via raising exceptions.

**Comment by sven1977** [comment]
Hi @simonsays1980 , thanks for suggesting this enhancement. I'll leave this open for discussion for now. In general, we are currently more moving toward giving up more responsibility from RLlib to the external libraries used, like gymnasium, so we can simplify RLlib. The same goes for giving back responsibility to the users (e.g. see the deprecation of `horizon` and `no_done_at_end` flags that are now fully served inside the user's env, and not handled by RLlib anymore).
Also, our env checker that you mentioned is only for pre checks, so it does not (and I believe should not) check things that are going on "deep" into training time, like e.g. some weird actions causing some weird behavior down the road.

**Comment by simonsays1980** [comment]
Hi @sven1977, 

generally I consider this as a good process, as it takes away some complexity from RLlib. I had [here](https://github.com/ray-project/ray/pull/31431) prepared a PR that wraps the `BaseEnv` - as this is the most basic environment in RLlib. In case of the environment a user might be able to write such a wrapper for her very own environment (gymnasium or any simulator). 

In case of error tracing inside of RLlib it is up to now not possible to easily trace back errors (or behaviors) without changing the source code of RLlib. I myself spend much of my time doing this - especially in case of TensorFlow it is often not easy to find out which function is calculating e.g. the gradients or how to print values, while in static graph mode. In my opinion this cannot easily be handed over to the user as she does need to have more knowledge about the library than is needed for application of it. I have some ideas how to handle this, too, and I am happy to share it in another feature request.
